common/types/traits/iterator.go

```diff
@@ -34,3 +34,16 @@ type Iterator interface {
 	// Next returns the next element.
 	Next() ref.Val
 }
+
+// Foldable aggregate types support iteration over (key, value) or (index, value) pairs.
+type Foldable interface {
+	// Fold invokes the Folder.FoldEntry for all entries in the type
+	Fold(Folder)
+}
+
+// Folder performs a fold on a given entry and indicates whether to continue folding.
+type Folder interface {
+	// FoldEntry indicates the calling type and the (key, value) pair associated with the entry.
+	// If the output is true, continue folding. Otherwise, terminate the fold.
+	FoldEntry(t ref.Type, key, val any) bool
```

Comment:
What is the purpose of passing the type to the `FoldEntry`? is this something that is going to be used later?

Reply:
It has to do with whether the fold is happening within one or two argument comprehensions to I know which value to assign where; however, I could remove this and just maintain support for the single and dual arg comprehensions separately in later CLs (might need to for performance reasons anyway)

Reply:
Removed. PTAL